File: PerfTester-Xamarin/PerfTester-Xamarin/Helpers/CSVHelper.cs
using System;
using System.IO;
using Foundation;
using PerfTesterXamarin.Models;
using UIKit;

namespace PerfTesterXamarin.Helpers
{
    public class CSVHelper
    {
        static string defaultIP = "192.168.0.11:8080";
        public static void SaveTestResults(Test test)
        {
            GetIPFromAlert((ip) =>
            {
                var fileName = GetFileName(test);
                var parsedHeaders = ParseHeadersIntoString(test);
                var parsedData = ParseResultsIntoString(test);
                var headersUrl = String.Format("http://{0}/EXPORT_HEADER.php?name={1}&data={2}", defaultIP, fileName, parsedHeaders);
                var dataUrl = String.Format("http://{0}/EXPORT_DATA.php?name={1}&data={2}", defaultIP, fileName, parsedData);
                var headersUri = new Uri(headersUrl);
                var dataUri = new Uri(dataUrl);
                var headersSet = new NSUrl(headersUri.GetComponents(UriComponents.HttpRequestUrl, UriFormat.UriEscaped));
                var dataSet = new NSUrl(dataUri.GetComponents(UriComponents.HttpRequestUrl, UriFormat.UriEscaped));
                if (headersSet != null)
                {
                    NSUrlSession.SharedSession.CreateDataTask(new NSUrlRequest(headersSet), (data, response, error) =>
                    {
                        if (dataSet != null)
                        {
                            NSUrlSession.SharedSession.CreateDataTask(new NSUrlRequest(dataSet)).Resume();
                        }
                    }).Resume();
                }
            });

        }

        static string GetFileName(Test test)
        {
            return String.Format("{0}.csv", test.Title);
        }

        static String ParseHeadersIntoString(Test test)
        {
            String finString = "Variant";
            for (var i = 0; i < test.Parameters.Length; i++)
            {
                var variant = i < test.Parameters.Length ? test.Parameters[i] : 0;
                finString += String.Format(",{0}", variant);
            }
            finString += String.Format("%0A");
            return finString;
        }

        static String ParseResultsIntoString(Test test)
        {
            String finString = "Visual Studio - ";
            var buildType = "REL";
#if DEBUG
            buildType = "DBG";
#endif
            finString += buildType;
            for (var i = 0; i < test.Parameters.Length; i++)
            {
                var result = i < test.Results.Length ? test.Results[i] : 0;
                finString += String.Format(",{0}", result);
            }
            finString += String.Format("%0A");
            return finString;
        }

        static void GetIPFromAlert(Action<String> completion)
        {
            UIAlertController alert = UIAlertController.Create("Where to save?", "Insert IP address to export data", UIAlertControllerStyle.Alert);
            UIAlertAction okButton = UIAlertAction.Create("Save", UIAlertActionStyle.Default, (obj) =>
            {
                completion(alert.TextFields[0].Text);
            });
            UIAlertAction cancelButton = UIAlertAction.Create("Cancel", UIAlertActionStyle.Destructive, null);
            alert.AddTextField( (tF) => 
            {
                tF.Text = defaultIP;
            });
            alert.AddAction(okButton);
            alert.AddAction(cancelButton);
            UIApplication.SharedApplication.Delegate.GetWindow().RootViewController.PresentViewController(alert, true, null);
        }
    }
}

File: PerfTester-Xcode/PerfTester-Xcode/Helpers/CSVHelper.swift
//
//  CSVHelper.swift
//  PerfTester-Xcode
//
//  Created by Patryk Średziński on 23.06.2018.
//  Copyright © 2018 private.perftesterxcode. All rights reserved.
//

import UIKit

class CSVHelper {

    static let defaultIP = "10.12.141.26:8080"

    static func saveTestResults(test: Test) {
        getIPFromAlert() { ip in
            let fileName = getFileName(test: test)
            let parsedHeaders = parseHeadersIntoString(test: test)
            let parsedData = parseResultsIntoString(test: test)
            let headersUrl = "http://\(ip)/EXPORT_HEADER.php?name=\(fileName)&data=\(parsedHeaders)"
            let dataUrl = "http://\(ip)/EXPORT_DATA.php?name=\(fileName)&data=\(parsedData)"
            let encodedHeadersUrl = headersUrl.addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed)
            let encodedDataUrl = dataUrl.addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed)
            if let urlSet = URL(string: encodedHeadersUrl ?? "") {
                URLSession.shared.dataTask(with: urlSet).resume()
                URLSession.shared.dataTask(with: urlSet, completionHandler: { _,_,_ in
                    if let urlSet = URL(string: encodedDataUrl ?? "") {
                        URLSession.shared.dataTask(with: urlSet).resume()
                    }
                }).resume()
            }
        }
    }
    
    fileprivate static func getFileName(test: Test) -> String {
        return String(format: "%@.csv", test.title)
    }
    
    fileprivate static func parseHeadersIntoString(test: Test) -> String {
        var finString = "Variant"
        for i in 0...test.parameters.count-1 {
            let variant = i < test.parameters.count ? test.parameters[i] : 0
            finString = finString.appendingFormat(",%d", Int(round(variant)))
        }
        finString = finString.appending("\n")
        return finString
    }
    
    fileprivate static func parseResultsIntoString(test: Test) -> String {
        var finString = "Xcode - "
        var buildType = "REL"
        #if DEBUG
            buildType = "DBG"
        #endif
        finString = finString.appending(buildType)
        for i in 0...test.parameters.count-1 {
            let result = i < test.results.count ? test.results[i] : 0
            finString = finString.appendingFormat(",%d", Int(round(result)))
        }
        finString = finString.appending("\n")
        return finString
    }
    
    fileprivate static func getIPFromAlert(completion: @escaping (String) -> Void) {
        let alert = UIAlertController(title: "Where to save?", message: "Insert IP address to export data", preferredStyle: .alert)
        let okButton = UIAlertAction(title: "Save", style: .default) { _ in
            completion(alert.textFields?.first?.text ?? defaultIP)
        }
        let cancelButton = UIAlertAction(title: "Cancel", style: .destructive, handler: nil)
        alert.addTextField { tF in
            tF.text = defaultIP
        }
        alert.addAction(okButton)
        alert.addAction(cancelButton)
        UIApplication.shared.delegate?.window??.rootViewController?.present(alert, animated: true, completion: nil)
    }
}
